forte/processors/data_augment/algorithms/UDA.py

```diff
@@ -0,0 +1,192 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file implements the UDA(Unsupervised Data Augmentation)
+algorithm as described in
+Unsupervised Data Augmentation for Consistency Training
+(https://arxiv.org/abs/1904.12848)
+(https://github.com/google-research/uda)
+"""
+from typing import Optional, List, Callable, Tuple, Iterator
+from torch import Tensor
+from texar.torch.losses.info_loss import kl_divg_loss_with_logits
+from texar.torch.data import DataIterator
+from texar.torch.data.data.dataset_utils import Batch
+
+
+class UDAIterator:
+    r"""
+    This iterator wraps the Unsupervised Data Augmentation(UDA)
+    algorithm by calculating the unsupervised loss automatically
+    during each iteration. It takes both supervised and unsupervised
+    data iterator as input.
+
+    The unsupervised data should contain the original input and the
+    augmented input. The original and augmented inputs should be in
+    the same training example. The user should provide a function
+    unsup_forward_fn to calculate the output logits for original &
+    augmented inputs from the unsupervised batch.
+
+    During each iteration, the iterator will calculate the UDA loss
+    automatically with the given unsup_forward_fn. User can combine the
+    UDA loss with the supervised loss for model training.
+
+    It uses tricks such as prediction sharpening and confidence masking.
+    Please refer to the UDA paper for more details.
+    (https://arxiv.org/abs/1904.12848)
+
+    Args:
+        - sup_iterator: The iterator for supervised data.
+        - unsup_iterator:
+            The iterator for unsupervised data. Each example in it
+            should contain both the original and augmented data.
+        - unsup_forward_fn:
+            The function to calculate the UDA loss. The signature should be:
+                unsup_forward_fn(unsup_batch: Batch) -> Tuple[Tensor, Tensor]
+            The return values are [original_logits, augmented_logits].
+            The last dimension of the logits should be the distribution of
+            prediction classes.
+        - softmax_temperature:
+            The softmax temperature for sharpening the distribution.
+        - confidence_threshold:
+            The threshold for confidence-masking. It is a threshold
+            of the probability in [0, 1], rather than of the logit.
+            If set to -1, the threshold will be ignored.
+        - reduction:
+            Default: 'mean'.
+            This is the same as the `reduction` argument in :func:
+            texar.losses.info_loss.kl_divg_loss_with_logits.
```

Comment:
Can you check the docstring to make sure user can click and find the corresponding function from here?